src/compiler/checker.ts

```diff
@@ -32137,7 +32141,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                         // outer call expression. Effectively we just want a snapshot of whatever has been
                         // inferred for any outer call expression so far.
                         const outerMapper = getMapperFromContext(cloneInferenceContext(outerContext, InferenceFlags.NoDefault));
-                        const instantiatedType = instantiateType(contextualType, outerMapper);
+                        const argumentApparentType = outerContext && isCallOrNewExpression(node.parent) && getDiscriminatedApparentType(node, contextualType);
+                        const instantiatedType = instantiateType(argumentApparentType && argumentApparentType !== unknownType ? argumentApparentType : contextualType, outerMapper);

```

Comment:
This is a different take on the problem that the one in https://github.com/microsoft/TypeScript/pull/52865 .

The `argumentApparentType !== unknownType` is really meant to be `isNotDefaultConstraintOfUnconstrainedTypeParamere(argumentApparentType)`. I thought that something like this would already exist but I couldn't find it. it's likely that without the strict null checks, we'd have to check against `emptyObjectType`. That's also why I temporarily added `@strict` here:
https://github.com/microsoft/TypeScript/pull/52866/files#diff-0ed2dcde3e5999198b7eebb67d0c508dbd4497536f956829f74706f33eca470dR3